[profiler] Support top-level memory events

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51421 [profiler] Support top-level memory events**

Summary:
Mark memory events that did not happen within an operator context
explicitly in the profiler output.

Test Plan:
python test/test_profiler.py -k test_memory_profiler
```
------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
              Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg       CPU Mem  Self CPU Mem      CUDA Mem  Self CUDA Mem    # of Calls
------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
        aten::rand        40.00%      10.000us       100.00%      25.000us      25.000us         400 b           0 b           0 b           0 b             1
       aten::empty        24.00%       6.000us        24.00%       6.000us       6.000us         400 b         400 b           0 b           0 b             1
    aten::uniform_        36.00%       9.000us        36.00%       9.000us       9.000us           0 b           0 b           0 b           0 b             1
          [memory]         0.00%       0.000us         0.00%       0.000us       0.000us        -400 b        -400 b        -512 b        -512 b             2
------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
```

Differential Revision: [D26166518](https://our.internmc.facebook.com/intern/diff/D26166518)

Commit:
[profiler] Support top-level memory events

Summary:
Mark memory events that did not happen within an operator context
explicitly in the profiler output.
This PR also adds an API to track memory events outside of or partially
overlapping with the profiler scope.

Test Plan:
python test/test_profiler.py -k test_memory_profiler

[ghstack-poisoned]

ngimel commented on c10/core/CPUAllocator.cpp:
```diff
@@ -292,6 +292,10 @@ void ProfiledCPUMemoryReporter::Delete(void* ptr) {
       allocated = allocated_;
       nbytes = it->second;
       size_table_.erase(it);
+    } else {
```
are there any changes required in CUDAAllocator?

ilia-cher replied:
CUDACachingAllocator already saves block sizes

Commit:
Update on "[profiler] Support top-level memory events"


Summary:
Mark memory events that did not happen within an operator context
explicitly in the profiler output.
This PR also adds an API to track memory events outside of or partially
overlapping with the profiler scope.

Test Plan:
python test/test_profiler.py -k test_memory_profiler
```
------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
              Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg       CPU Mem  Self CPU Mem      CUDA Mem  Self CUDA Mem    # of Calls
------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
        aten::rand        40.00%      10.000us       100.00%      25.000us      25.000us         400 b           0 b           0 b           0 b             1
       aten::empty        24.00%       6.000us        24.00%       6.000us       6.000us         400 b         400 b           0 b           0 b             1
    aten::uniform_        36.00%       9.000us        36.00%       9.000us       9.000us           0 b           0 b           0 b           0 b             1
          [memory]         0.00%       0.000us         0.00%       0.000us       0.000us        -400 b        -400 b        -512 b        -512 b             2
------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
```

Differential Revision: [D26166518](https://our.internmc.facebook.com/intern/diff/D26166518)

[ghstack-poisoned]

ilia-cher commented:
(we decided not to include the new api at the moment, will remove enable_global_memory_reporting)

Commit:
Update on "[profiler] Support top-level memory events"


Summary:
Mark memory events that did not happen within an operator context
explicitly in the profiler output.
This PR also adds an API to track memory events outside of or partially
overlapping with the profiler scope.

Test Plan:
python test/test_profiler.py -k test_memory_profiler
```
------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
              Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg       CPU Mem  Self CPU Mem      CUDA Mem  Self CUDA Mem    # of Calls
------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
        aten::rand        40.00%      10.000us       100.00%      25.000us      25.000us         400 b           0 b           0 b           0 b             1
       aten::empty        24.00%       6.000us        24.00%       6.000us       6.000us         400 b         400 b           0 b           0 b             1
    aten::uniform_        36.00%       9.000us        36.00%       9.000us       9.000us           0 b           0 b           0 b           0 b             1
          [memory]         0.00%       0.000us         0.00%       0.000us       0.000us        -400 b        -400 b        -512 b        -512 b             2
------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
```

Differential Revision: [D26166518](https://our.internmc.facebook.com/intern/diff/D26166518)

[ghstack-poisoned]

Commit:
Update on "[profiler] Support top-level memory events"


Summary:
Mark memory events that did not happen within an operator context
explicitly in the profiler output.
This PR also adds an API to track memory events outside of or partially
overlapping with the profiler scope.

Test Plan:
python test/test_profiler.py -k test_memory_profiler
```
------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
              Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg       CPU Mem  Self CPU Mem      CUDA Mem  Self CUDA Mem    # of Calls
------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
        aten::rand        40.00%      10.000us       100.00%      25.000us      25.000us         400 b           0 b           0 b           0 b             1
       aten::empty        24.00%       6.000us        24.00%       6.000us       6.000us         400 b         400 b           0 b           0 b             1
    aten::uniform_        36.00%       9.000us        36.00%       9.000us       9.000us           0 b           0 b           0 b           0 b             1
          [memory]         0.00%       0.000us         0.00%       0.000us       0.000us        -400 b        -400 b        -512 b        -512 b             2
------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
```

Differential Revision: [D26166518](https://our.internmc.facebook.com/intern/diff/D26166518)

[ghstack-poisoned]

ngimel commented on test/test_profiler.py:
```diff
@@ -163,6 +165,129 @@ def test_kineto_multigpu(self):
+    def test_memory_profiler(self):
+        def run_profiler(tensor_creation_fn, metric):
+            # collecting allocs / deallocs
+            with _profile(profile_memory=True, record_shapes=True, use_kineto=kineto_available()) as prof:
```
we should make a follow-up item to either migrate or duplicate the tests with `profiler` instead of `autograd.profiler`.

ilia-cher replied:
all our autograd.profiler tests should use `use_kineto=kineto_available()` now, and we have both kineto-enabled and not enabled CI builds; eventually when autograd.profiler is deprecated and we always use kineto, we will just make the tests use the new, kineto-only api

ngimel commented on test/test_profiler.py:
```diff
@@ -163,6 +165,129 @@ def test_kineto_multigpu(self):
+                del y
+            gc.collect()
+        stats = prof.key_averages(group_by_input_shape=True)
+        print(stats.table(sort_by="cpu_memory_usage", row_limit=-1))
```
this looks like a debug print

ngimel commented on test/test_profiler.py:
```diff
@@ -163,6 +165,129 @@ def test_kineto_multigpu(self):
+        def create_mkldnn_tensor():
+            return torch.rand(10, 10, dtype=torch.float32).to_mkldnn()
+
+        print("Running CPU test")
```
remove debug print

ilia-cher replied:
it was originally added on purpose (together with profiler output), I think we have many tests that do this?

ngimel replied:
Idk, usually tests don't do this (they are tests), but there are some stray prints. Do you need this output?

ilia-cher replied:
only for debug, will remove

ngimel reviewed (approved):
Please remove debug prints.

Commit:
Update on "[profiler] Support top-level memory events"


Summary:
Mark memory events that did not happen within an operator context
explicitly in the profiler output.

Test Plan:
python test/test_profiler.py -k test_memory_profiler
```
------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
              Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg       CPU Mem  Self CPU Mem      CUDA Mem  Self CUDA Mem    # of Calls
------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
        aten::rand        40.00%      10.000us       100.00%      25.000us      25.000us         400 b           0 b           0 b           0 b             1
       aten::empty        24.00%       6.000us        24.00%       6.000us       6.000us         400 b         400 b           0 b           0 b             1
    aten::uniform_        36.00%       9.000us        36.00%       9.000us       9.000us           0 b           0 b           0 b           0 b             1
          [memory]         0.00%       0.000us         0.00%       0.000us       0.000us        -400 b        -400 b        -512 b        -512 b             2
------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------
```

Differential Revision: [D26166518](https://our.internmc.facebook.com/intern/diff/D26166518)

[ghstack-poisoned]
